### ISSUE TITLE
use scikit-learn check_estimator to ensure compatibiity of Nilearn estimators with scikit-learn

### ISSUE THREAD

**Issue by Remi-Gau** [nilearn/nilearn#4538]
see https://scikit-learn.org/dev/developers/develop.html#rolling-your-own-estimator

quickly checked on a couple of Nilearn estimators and they seem to fail: will add code examples when in front of a real computer

may be worth discussing to see if we want to have 100% compliance or if we should have our own version of check_estimator

**Comment by Remi-Gau** [comment]
Yup me too.
Will list the errors we get in this issue so we can  make a proper decision.

**Comment by Remi-Gau** [comment]
one example that loops through all the checks

```python
from sklearn.utils.estimator_checks import check_estimator
from nilearn.decoding.decoder import (
    Decoder,
)

def test_check_estimator():
    """Check compliance with sklearn estimators."""
    model = Decoder()

    for estimator, check in check_estimator(estimator=model, generate_only=True):
        try:
            check(estimator)
        except Exception:
            print(f"FAIL: {check}")
        else:
            print(f"SUCCESS: {check}")
```

```
SUCCESS: functools.partial(<function check_no_attributes_set_in_init at 0x735a695ef240>, 'Decoder')
FAIL: functools.partial(<function check_estimators_dtypes at 0x735a695eca40>, 'Decoder')
FAIL: functools.partial(<function check_fit_score_takes_y at 0x735a695ec900>, 'Decoder')
FAIL: functools.partial(<function check_estimators_fit_returns_self at 0x735a695ee340>, 'Decoder')
FAIL: functools.partial(<function check_estimators_fit_returns_self at 0x735a695ee340>, 'Decoder', readonly_memmap=True)
FAIL: functools.partial(<function check_complex_data at 0x735a695e3420>, 'Decoder')
FAIL: functools.partial(<function check_dtype_object at 0x735a695e3380>, 'Decoder')
FAIL: functools.partial(<function check_estimators_empty_data_messages at 0x735a695ecc20>, 'Decoder')
FAIL: functools.partial(<function check_pipeline_consistency at 0x735a695ec7c0>, 'Decoder')
FAIL: functools.partial(<function check_estimators_nan_inf at 0x735a695ecd60>, 'Decoder')
FAIL: functools.partial(<function check_estimators_overwrite_params at 0x735a695ef100>, 'Decoder')
FAIL: functools.partial(<function check_estimator_sparse_array at 0x735a695e2b60>, 'Decoder')
FAIL: functools.partial(<function check_estimator_sparse_matrix at 0x735a695e2ac0>, 'Decoder')
FAIL: functools.partial(<function check_estimators_pickle at 0x735a695ecfe0>, 'Decoder')
FAIL: functools.partial(<function check_estimators_pickle at 0x735a695ecfe0>, 'Decoder', readonly_memmap=True)
SUCCESS: functools.partial(<function check_estimator_get_tags_default_keys at 0x735a695f4680>, 'Decoder')
FAIL: functools.partial(<function check_classifier_data_not_an_array at 0x735a695ef4c0>, 'Decoder')
FAIL: functools.partial(<function check_classifiers_one_label at 0x735a695ed760>, 'Decoder')
SUCCESS: functools.partial(<function check_classifiers_one_label_sample_weights at 0x735a695ed8a0>, 'Decoder')
FAIL: functools.partial(<function check_classifiers_classes at 0x735a695ee840>, 'Decoder')
SUCCESS: functools.partial(<function check_estimators_partial_fit_n_features at 0x735a695ed120>, 'Decoder')
FAIL: functools.partial(<function check_classifiers_train at 0x735a695ed9e0>, 'Decoder')
FAIL: functools.partial(<function check_classifiers_train at 0x735a695ed9e0>, 'Decoder', readonly_memmap=True)
FAIL: functools.partial(<function check_classifiers_train at 0x735a695ed9e0>, 'Decoder', readonly_memmap=True, X_dtype='float32')
FAIL: functools.partial(<function check_classifiers_regression_target at 0x735a695eff60>, 'Decoder')
FAIL: functools.partial(<function check_supervised_y_no_nan at 0x735a695e1bc0>, 'Decoder')
FAIL: functools.partial(<function check_supervised_y_2d at 0x735a695ee5c0>, 'Decoder')
FAIL: functools.partial(<function check_estimators_unfitted at 0x735a695ee480>, 'Decoder')
SUCCESS: functools.partial(<function check_non_transformer_estimators_n_iter at 0x735a695efa60>, 'Decoder')
SUCCESS: functools.partial(<function check_decision_proba_consistency at 0x735a695f40e0>, 'Decoder')
SUCCESS: functools.partial(<function check_parameters_default_constructible at 0x735a695ef7e0>, 'Decoder')
FAIL: functools.partial(<function check_methods_sample_order_invariance at 0x735a695e3ba0>, 'Decoder')
FAIL: functools.partial(<function check_methods_subset_invariance at 0x735a695e3a60>, 'Decoder')
FAIL: functools.partial(<function check_fit2d_1sample at 0x735a695e3d80>, 'Decoder')
FAIL: functools.partial(<function check_fit2d_1feature at 0x735a695e3f60>, 'Decoder')
SUCCESS: functools.partial(<function check_get_params_invariance at 0x735a695efce0>, 'Decoder')
SUCCESS: functools.partial(<function check_set_params at 0x735a695efe20>, 'Decoder')
FAIL: functools.partial(<function check_dict_unchanged at 0x735a695e3560>, 'Decoder')
FAIL: functools.partial(<function check_dont_overwrite_parameters at 0x735a695e3740>, 'Decoder')
FAIL: functools.partial(<function check_fit_idempotent at 0x735a695f42c0>, 'Decoder')
FAIL: functools.partial(<function check_fit_check_is_fitted at 0x735a695f4360>, 'Decoder')
FAIL: functools.partial(<function check_n_features_in at 0x735a695f4400>, 'Decoder')
FAIL: functools.partial(<function check_fit1d at 0x735a695ec0e0>, 'Decoder')
FAIL: functools.partial(<function check_fit2d_predict1d at 0x735a695e3880>, 'Decoder')
```

From a quick inspection it looks like most of the checks that fail actually do so because  the `fit()` method in scikit-learn is tested with numpy arrays whereas we expect most of the time list of nifti or surface images.

**Comment by bthirion** [comment]
Indeed, we don't want to enforce Numpy arrays.
So the long-term solution is to have our own checks.

**Comment by Remi-Gau** [comment]
yeah I am afraid so.

this may take some time but given there are several checks we can probably add them one by one through several PRs

**Comment by Remi-Gau** [comment]
follow up on this:
- [ ] identify which check fails or pass for each of our estimator
- [ ] identify which estimator should always fail for each estimator (for example do not fit numpy array)
- [ ] identify which checks that currently fails that we care about fixing (for example pipeline compatibility) and possibly fix them
- [ ] see if we want / can create some of our own checks

**Comment by Remi-Gau** [comment]
Maybe worth double checking if those tests are related to this issue: https://github.com/nilearn/nilearn/blob/main/nilearn/decoding/tests/test_sklearn_compatibility.py

Or if those tests are still needed.

**Comment by Remi-Gau** [comment]
One thing we should probably do rather soon is create our own 'tags' and have some 'input tag' (see [sklearn ones](https://scikit-learn.org/stable/modules/generated/sklearn.utils.InputTags.html)) that says input is niimg-like, so we can more easily skip some of the sklearn checks that assume arrays.

**Comment by Remi-Gau** [comment]
updating top message of this issue with a list of sklearn checks, which ones are implemented, which ones are skipped for maskers...

**Comment by Remi-Gau** [comment]
next steps on this:

# TL;DR

We now have more checks that pass than fail, so keeping track of the expected failures will be easier than keeping track of all the successes.

## rely on estimator_check generator from sklearn and its 'blocklist'

We were mostly keeping track of the sklearn checks that were **valid** for our estimators. When a check is not in that list, it is assumed it will fail (so by default any new check added upstream in sklearn would fail).

### test_check_estimator_valid

- valid check pass -> test pass
- valid check fails -> test fails

### test_check_estimator_invalid

- invalid check fails -> test pass
- invalid check pass -> no error when failure expected -> test fails (as we run our tests in xfail strict mode)

### original motivation

This approach was useful at the beginning because we had a lot more failures than success on those checks. Also we did not know which of the sklearn failed for a good reason or just failed because the check did not make sense for some of our estimators (for example sklearn estimators expect np.array as input for fit whereas our estimators often work with nifti object).

### sklearn 1.6

Since sklearn 1.6, their estimator_check generator includes an expected_failure parameter to list all the checks (for a given estimator) that are expected to fail and the reason for failing: it acts as a 'blocklist'.
So our 'valid list' is kinda duplicating this.

Also now it is becoming easier to keep track of failures than successes: fewer failures.

Keeping track of a blocklist is also more lenient: if a new check is added upstream, it won't be assumed that it will fail.

## actions

I have started removing the extra_valid_checks that existed for our different estimators:
- https://github.com/nilearn/nilearn/pull/5229
- https://github.com/nilearn/nilearn/pull/5236
- they should all be removed by #5246 

Then we should remove the global list of valid and adapt the code to only rely on expected_failed_checks.

https://github.com/nilearn/nilearn/blob/31d4669cf7b8a38f58be0c263077b77881cff1b8/nilearn/_utils/estimator_checks.py#L54-L87

This should also switching the behavior of the sklearn check generator from "skip" (what we do currently):

https://github.com/nilearn/nilearn/blob/31d4669cf7b8a38f58be0c263077b77881cff1b8/nilearn/_utils/estimator_checks.py#L237-L239


to using "xfail" to actually make sure the tests that are expected to fail indeed fail. Note this will probably increase the run time of the test suite a bit.

**Comment by Remi-Gau** [comment]
>This should also switching the behavior of the sklearn check generator from "skip" (what we do currently):
to using "xfail" to actually make sure the tests that are expected to fail indeed fail. Note this will probably increase the run time of the test suite a bit.

This will have to wait a bit as there seems to be something a bit off on the sklearn side: https://github.com/scikit-learn/scikit-learn/issues/31408

**Comment by Remi-Gau** [comment]
> we should remove the global list of valid and adapt the code to only rely on expected_failed_checks.

I think this will be easier to do once we drop support for sklearn 1.5 so that's gonna take a while

**Comment by Remi-Gau** [comment]
Will create separate issues for the different types of nilearn estimators to better track their compliance with scikit learn requirements.

**Comment by Remi-Gau** [comment]
replaced by issues specific to each category of estimator
